Player: Modified reputation gain calculations

JJBcku opened the pull request:
I have modified the calculation of level difference penalties for reputation gain to be more in line with the original World of Warcraft.

## 🍰 Pullrequest
<!-- Describe the Pullrequest. -->

### Proof
<!-- Link resources as proof -->
- [Reddit](https://www.reddit.com/r/classicwow/comments/akoomj/reputation_gains_in_vanilla_wow_did_they_diminish/)
- [Reddit 2](https://www.reddit.com/r/classicwow/comments/bcwmfl/is_it_true_that_in_vanilla_quests_give_less_rep/)
- [Blizzard forum](https://us.forums.blizzard.com/en/wow/t/wow-classic-faction-reputation-gained-incorrect/948304)

### Issues
<!-- Which Issues does this fix, which are related?
- fixes #XXX
- relates #XXX
-->
- fixes non vanilla wow-like reputation gains

JJBcku commented:
Could someone tell me why there is no discussion or even acknowledgment of my pull request?
Have I did something wrong?

insunaa commented:
Currently nobody on the dev team has the time to review pull requests

killerwife commented:
@brotalnia any input on this?

killerwife commented:
@ratkosrb

killerwife commented on src/game/Entities/Player.cpp:
```diff
@@ -6282,15 +6282,26 @@ int32 Player::CalculateReputationGain(ReputationSource source, int32 rep, int32
-        if (currentLevel > treshold)
-            percent *= std::max(minRate, (1.0f - (0.2f * (currentLevel - treshold))));
+            if (lastGreenLevel > creatureOrQuestLevel)
+                percent *= std::max(minRate, (1.0f - (0.2f * (lastGreenLevel - creatureOrQuestLevel))));
```
I understand this is based on the forum post, but have you actually tested this on classic PTR or what specifically is this based on (I have read the links pointing it out)

killerwife replied:
After careful investigation, the function GetGrayLevel is deprecated and IsTrivialLevelDifference is the correct one. (trivial means grey in official naming)

I went ahead and a 48 npc (green) on lvl 60 gives full rep, and the GetGrayLevel function would not do that properly

killerwife commented:
My proposed code based on client lua:

`
    if (source == REPUTATION_SOURCE_KILL)
    {
        if (MaNGOS::XP::IsTrivialLevelDifference(currentLevel, creatureOrQuestLevel))
        {
            const uint32 greenRange = MaNGOS::XP::GetQuestGreenRange(currentLevel);
            percent *= std::max(minRate, (1.0f - (0.2f * (currentLevel - greenRange - creatureOrQuestLevel))));
        }
    }
    else if (source == REPUTATION_SOURCE_QUEST)
    {
        // Pre-3.0.8: Declines with 20% for each level if 6 levels or more below the player down to a minimum (default: 20%)
        const uint32 treshold = (creatureOrQuestLevel + 5);

        if (currentLevel > treshold)
            percent *= std::max(minRate, (1.0f - (0.2f * (currentLevel - treshold))));
    }
`

Test case, grey mob in vanilla at 47 (grey diff in vanilla is 12 levels per client and lua) will get 80% rep. Quest case unchanged.

JJBcku commented:
Hi,
I have added checking for the trivial level difference in the modified function.
I have not found a function named GetQuestGreenRange in Mangos::XP so instead I have modified the GetGrayLevel function to return values in line with the ones IsTrivialLevelDifference is using.

killerwife commented:
The function exists in lua client code. I extracted it over here: https://github.com/cmangos/mangos-classic/commit/9e95f5d618bc35ebe7a17d93518dc9d34416c805 take a look

JJBcku commented:
Thank you for your help.
Could you tell me the email I should use to credit you as a co-author.

killerwife commented:
Just look at git history in commandline, its there. I prefer not to write it here cos it kinda leads to more spam email for me :D

JJBcku commented:
I have added the changes you suggested.
I have kept in the fix for the GetGrayLevel function, is that okay?

killerwife commented:
I will review how all 3 cores use them and then I will merge it. We prefer to use official functions from client whenever named and available. The gray level one doesnt exist in client.
